examples/dataflow-scala-kafka2avro/src/main/scala/Kafka2Avro.scala

```diff
@@ -0,0 +1,168 @@
+    */
+  def windowIn(ps: SCollection[T]): SCollection[(IntervalWindow, Iterable[T])] = {
+    val windows: SCollection[(IntervalWindow, Iterable[T])] =
+      ps
+      // Here we just make fixed windows of 30 secs, as an example
+      .withFixedWindows(Duration.standardSeconds(30))
+      .withWindow[IntervalWindow]
+      .swap
+      .groupByKey
+
+    windows
+  }
+
+  /** The main function of this application
+    *
+    * @param args array of command line arguments
+    */
+  def main(cmdLineArgs: Array[String]): Unit = {
+    val myConfigEither: Either[ConfigReaderFailures, Kafka2AvroConfig] =
+      pureconfig.loadConfig[Kafka2AvroConfig]
+
+    myConfigEither match {
+      case Left(configErrors: ConfigReaderFailures) =>
+        Kafka2AvroUtils.failAndExitWithWrongConfig(configErrors, logger)
+      case Right(configOk: Kafka2AvroConfig) =>
+        implicit val config = configOk
+
+        implicit val (sc: ScioContext, _) = ContextAndArgs(cmdLineArgs)
+
+        // Use Kryo for serialization
+        implicit def coderKafkaRecord: Coder[T] = Coder.kryo[T]
+
+        // ETL process
+
+        val outputLocation = "gs://%s/%s".format(
+          config.destBucket,
+          config.destPath)
+
+        val extracted   = extract
+        val transformed = transform(extracted)
+        val _           = load(transformed, outputLocation)
+
+        sc.close
+    }
+  }
+
+
+  /** Read data from Kafka, and return the messages ignoring the keys */
+  def extract(implicit config: Kafka2AvroConfig, sc: ScioContext): SCollection[String] = {
+    val messages: SCollection[KV[String, String]] = sc.customInput("ReadFromKafka",
+      KafkaIO.read[String, String]
+        .withBootstrapServers(config.broker)
+        .withTopic(config.kafkaTopic)
+        .withKeyDeserializer(classOf[StringDeserializer])
+        .withValueDeserializer(classOf[StringDeserializer])
+        .withoutMetadata
+    )
+
+    // We will not use they key, so let's drop it to save some memory
+    messages.map(_.getValue)
+  }
+
+  /** Transform the strings read from Kafka
+    *
+    * Here we try to decode the string into an object, and then group in a
+    * window.
+    */
+  def transform(messages: SCollection[String]): SCollection[(IntervalWindow, Iterable[T])] = {
+    val records: SCollection[T] = messages.map(Kafka2AvroUtils.string2object[T])
+
+    // The windowIn function is assumed to be in scope.
+    // It could also be passed as an argument to the transform step
+    val groups: SCollection[(IntervalWindow, Iterable[T])] =
+      records.transform("Windowing")(windowIn)
+
+    groups
+  }
+
+  /** Load data into GCS
+    *
+    * After the transform step, we should now have a collection of objects,
+    * grouped in windows. We will write each window in a separate file in GCS,
+    * encoded as Avro.
+    */
+  def load(records: SCollection[(IntervalWindow, Iterable[T])], location: String)(implicit sc: ScioContext): Unit = {
```

Comment:
Looks like SCIO does have support for AvroIO so should we use that instead of manually writing Avro files to GCS?

Reply:
I was not aware of that possibility with SCIO. Let me see how I can include it. Thanks for the tip!

Reply:
I am not sure if AvroIO can be applied to windowed outputs. For what I have found, saveAsAvroFile is available for SCollection. But there is no way to apply it to a local collection (List, Iterable, etc.).

I am using the same method to write the Avro output as the SCIO example for a windowed version of WordCount (streaming pipeline). This example does not actually use TextIO to save the output: https://github.com/spotify/scio/blob/master/scio-examples/src/main/scala/com/spotify/scio/examples/WindowedWordCount.scala#L83-L91

So I will resolve now this comment, but please reopen it if I am missing something about how to generate the Avro output for windowed output.

Thanks!